File: lib-segment/src/main/java/websiteschema/mpsegment/filter/AbstractSegmentFilter.java
package websiteschema.mpsegment.filter;

import websiteschema.mpsegment.core.SegmentResult;

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractSegmentFilter implements ISegmentFilter {
    interface Operation {
        public void modify(SegmentResult segmentResult);
    }

    class DeleteOperation implements Operation {
        int index;
        DeleteOperation(int index) {
            this.index = index;
        }
        public void modify(SegmentResult segmentResult) {
            segmentResult.markWordToBeDeleted(index);
        }
    }

    class MergeOperation implements Operation {
        int start;
        int end;
        int pos;

        MergeOperation(int start, int end, int pos) {
            this.start = start;
            this.end = end;
            this.pos = pos;
        }

        public void modify(SegmentResult segmentResult) {
            segmentResult.merge(start, end, pos);
        }
    }

    protected SegmentResult segmentResult;
    private int wordPosIndexes[];
    private List<Operation> operationSettings = new ArrayList<Operation>();

    public abstract void doFilter();

    @Override
    public void filtering() {
        doFilter();
        compactSegmentResult();
    }

    @Override
    public void setSegmentResult(SegmentResult segmentResult) {
        this.segmentResult = segmentResult;
        wordPosIndexes = new int[segmentResult.length()];
    }

    protected void setWordIndexesAndPOSForMerge(int startWordIndex, int endWordIndex, int POS) {
        operationSettings.add(new MergeOperation(startWordIndex, endWordIndex, POS));
        markWordsHasBeenRecognized(startWordIndex, endWordIndex, POS);
    }

    protected void deleteWordAt(int index) {
        operationSettings.add(new DeleteOperation(index));
    }

    protected void compactSegmentResult() {
        if (operationSettings.size() > 0) {
            for (Operation mergeSetting : operationSettings) {
                mergeSetting.modify(segmentResult);
            }
            segmentResult.compact();
            operationSettings.clear();
        }
    }

    protected boolean isWordConfirmed(int wordIndex) {
        return wordPosIndexes[wordIndex] > 0;
    }

    protected boolean isNotMarked(int index) {
        return wordPosIndexes[index] <= 0;
    }

    private void markWordsHasBeenRecognized(int startWordIndex, int endWordIndex, int POS) {
        for (int i = startWordIndex; i <= endWordIndex; i++) {
            wordPosIndexes[i] = POS;
        }
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/filter/AbstractSegmentFilter.scala
package websiteschema.mpsegment.filter

import websiteschema.mpsegment.core.SegmentResult
import collection.mutable.ListBuffer
import websiteschema.mpsegment.dict.{POSUtil, DictionaryFactory}

abstract class AbstractSegmentFilter extends ISegmentFilter {

  trait Operation {
    def modify(segmentResult: SegmentResult)
  }

  class DeleteOperation(index: Int) extends Operation {

    def modify(segmentResult: SegmentResult) {
      segmentResult.markWordToBeDeleted(index)
    }
  }

  class MergeOperation(start: Int, end: Int, pos: Int) extends Operation {

    def modify(segmentResult: SegmentResult) {
      segmentResult.merge(start, end, pos)
    }
  }

  class SeparateOperation(index: Int, pos: Int) extends Operation {

    def modify(segmentResult: SegmentResult) {
      val rest = segmentResult(index).word.substring(1)
      val word = DictionaryFactory().getCoreDictionary().getWord(rest)
      if (word != null) {
        segmentResult.separate(index, 1, pos, word.getWordPOSTable()(0)(0))
      } else {
        segmentResult.separate(index, 1, pos, POSUtil.POS_UNKOWN)
      }
    }
  }

  var segmentResult: SegmentResult = null
  private var wordPosIndexes: Array[Int] = null
  private val operationSettings = ListBuffer[Operation]()

  def doFilter()

  override def filtering() {
    doFilter()
    compactSegmentResult()
  }

  override def setSegmentResult(segmentResult: SegmentResult) {
    this.segmentResult = segmentResult
    wordPosIndexes = new Array[Int](segmentResult.length())
  }

  def setWordIndexesAndPOSForMerge(startWordIndex: Int, endWordIndex: Int, POS: Int) {
    operationSettings += (new MergeOperation(startWordIndex, endWordIndex, POS))
    markWordsHasBeenRecognized(startWordIndex, endWordIndex, POS)
  }

  def deleteWordAt(index: Int) {
    operationSettings += (new DeleteOperation(index))
  }

  def separateWordAt(index: Int, partOfSpeech: Int) {
    operationSettings += (new SeparateOperation(index, partOfSpeech))
  }

  def compactSegmentResult() {
    if (operationSettings.size > 0) {
      for (mergeSetting <- operationSettings) {
        mergeSetting.modify(segmentResult)
      }
      segmentResult.compact()
      operationSettings.clear
    }
  }

  def isWordConfirmed(wordIndex: Int): Boolean = {
    return wordPosIndexes(wordIndex) > 0
  }

  def isNotMarked(index: Int): Boolean = {
    return wordPosIndexes(index) <= 0
  }

  private def markWordsHasBeenRecognized(startWordIndex: Int, endWordIndex: Int, POS: Int) {
    for (i <- startWordIndex to endWordIndex) {
      wordPosIndexes(i) = POS
    }
  }
}
